(ready for review) Add a Key Features page to docs

Steph-Chong opened the pull request:
Adds a "Key Features of `scores`" page to the docs. Resolves #568 

Very close (but not identical) to the key benefits section of the JOSS paper draft.

My logic for suggesting this for inclusion in the docs, is to make it easier for potential new users to assess if `scores` is likely to be of interest to them. 

I realise much (but not all) of this information is already in the README, but due to the need and desire to include other information in the README page, some of the key benefit information may not jump out as much as it once did. So I think this new page could be potentially useful. There is also some information in the key features page, that isn't in the README or elsewhere in the docs. For instance, I don't think the fact `scores` has been tested in HPC environments is stated anywhere in the docs currently. Possibly aside from the API, I don't think the docs mention `scores` handles missing data or masking of data (unless it has and I haven't understood that to be the case).

When I built it locally, it rendered correctly in readthedocs, and all links worked.

[Note: I initially called the page "key_benefits.md" and had the title "Key Benefits of `scores`". I have now updated this to "key_features.md" and retitled the page "Key Features of `scores`"]

Steph-Chong commented on docs/key_benefits.md:
```diff
@@ -0,0 +1,26 @@
+## Usability
+
+- A companion Jupyter Notebook [tutorial](project:./tutorials/Tutorial_Gallery.md) for each metric and statistical test that demonstrates its use in practice.
+- Novel scores not commonly found elsewhere (e.g., [ FIxed Risk Multicategorical (FIRM)](https://doi.org/10.1002/qj.4266), [Flip-Flop Index](https://doi.org/10.1002/met.1732)).
```
Re. Flip-Flop Index link:
- Ideally I would link to the entry in the "included.md" table, but I didn't know how to do that.
- When linking to the paper references, I know we normally link to both the original Flip-Flop Index article, and the Circular Flip-Flop Index article - but I couldn't do that in a single link. Is it okay to just link to the Flip-Flop Index article (as is currently the case), or does anyone have an alternative suggestion?
- Or, should it instead link to the API entry?

FIRM link:
- Ideally I would link to the entry in the "included.md" table, but I didn't know how to do that.
- Should this link to the paper? 
- Or should it link to the API?

Steph-Chong commented on docs/key_benefits.md:
```diff
@@ -0,0 +1,26 @@
+
+- A companion Jupyter Notebook [tutorial](project:./tutorials/Tutorial_Gallery.md) for each metric and statistical test that demonstrates its use in practice.
+- Novel scores not commonly found elsewhere (e.g., [ FIxed Risk Multicategorical (FIRM)](https://doi.org/10.1002/qj.4266), [Flip-Flop Index](https://doi.org/10.1002/met.1732)).
+- Complex scores (e.g., threshold-weighted continuous ranked probability score (twCRPS))
```
I added this section (this dot point isn't in the JOSS paper draft), as I thought it might be useful to include here. But, if people think this section should be shorter, I am also happy to remove this dot point.

Steph-Chong replied:
If this dot point stays: @nicholasloveday is twCRPS (a) the correct abbreviation and (b) is it useful to include it here? If I should either amend it, or remove this abbreviation, please let me know.

Steph-Chong commented on docs/key_benefits.md:
```diff
@@ -0,0 +1,26 @@
+- A companion Jupyter Notebook [tutorial](project:./tutorials/Tutorial_Gallery.md) for each metric and statistical test that demonstrates its use in practice.
+- Novel scores not commonly found elsewhere (e.g., [ FIxed Risk Multicategorical (FIRM)](https://doi.org/10.1002/qj.4266), [Flip-Flop Index](https://doi.org/10.1002/met.1732)).
+- Complex scores (e.g., threshold-weighted continuous ranked probability score (twCRPS))
+- Commonly-used scores are also included, so that `scores` can be used as a standalone package.
```
I modified this text from the JOSS paper draft - which says "Commonly-used scores are also included, meeting user requests to use scores as a standalone package." If anyone would prefer me to revert it to the same wording as the JOSS paper draft, let me know.

Steph-Chong commented on docs/key_benefits.md:
```diff
@@ -0,0 +1,26 @@
+## Usability
+
+- A companion Jupyter Notebook [tutorial](project:./tutorials/Tutorial_Gallery.md) for each metric and statistical test that demonstrates its use in practice.
+- Novel scores not commonly found elsewhere (e.g., [FIxed Risk Multicategorical (FIRM) score](https://doi.org/10.1002/qj.4266), [Flip-Flop Index](https://doi.org/10.1002/met.1732)).
```
Re. Flip-Flop Index link:
- Ideally I would link to the entry in the "included.md" table, but I didn't know how to do that.
- When linking to the paper references, I know we normally link to both the original Flip-Flop Index article, and the Circular Flip-Flop Index article - but I couldn't do that in a single link. Is it okay to just link to the Flip-Flop Index article (as is currently the case), or does anyone have an alternative suggestion?
- Or, should it instead link to the API entry?

FIRM link:
- Ideally I would link to the entry in the "included.md" table, but I didn't know how to do that.
- Should this link to the paper?
- Or should it link to the API?

Steph-Chong replied:
@nicholasloveday can I confirm the correct term for FIRM is FIxed Risk Multicategorical (FIRM) Score? 

I can't access the  https://doi.org/10.1002/qj.4266 article, as it is paywalled. 

But when I looked at the second version of the paper on arXiv (https://arxiv.org/pdf/2108.12814) - which I realise may not be identical to the final published version - there were two different descriptions of what FIRM stood for.
- On pages 3 and 7, it said "FIxed Risk Multicategory (FIRM) Forecast Framework"
- On page 16, it said "FIxed Risk Multicategorical (FIRM) framework"

So I just wanted to check that "FIxed Risk Multicategorical (FIRM) score" is correct.

tennlee commented:
I haven't considered my views on this. I have pushed the changes up to branch 228 so they can be visualised at https://scores.readthedocs.io/en/228-documentation-testing-branch/

Steph-Chong commented on docs/key_features.md:
```diff
@@ -0,0 +1,26 @@
+
+- A companion Jupyter Notebook [tutorial](project:./tutorials/Tutorial_Gallery.md) for each metric and statistical test that demonstrates its use in practice.
+- Novel scores not commonly found elsewhere (e.g., [FIxed Risk Multicategorical (FIRM) score](https://doi.org/10.1002/qj.4266), [Flip-Flop Index](https://doi.org/10.1002/met.1732)).
+- Complex scores (e.g., threshold-weighted continuous ranked probability score (twCRPS))
```
I added this dot point about complex scores (this dot point isn't in the JOSS paper draft), as I thought it might be useful to include here. But, if people think this section should be shorter, I am also happy to remove this dot point.

Steph-Chong replied:
@nicholasloveday is "twCRPS" (a) the correct abbreviation and (b) is it useful to include it here? If I should either amend it, or remove this abbreviation, please let me know.

Steph-Chong replied:
Note for self: add a full stop at the end of this dot point.

nicholasloveday replied:
Yes, "twCRPS" is the correct abbreviation. I don't mind whether that dot point is included or not.

Steph-Chong commented on docs/key_features.md:
```diff
@@ -0,0 +1,26 @@
+- A companion Jupyter Notebook [tutorial](project:./tutorials/Tutorial_Gallery.md) for each metric and statistical test that demonstrates its use in practice.
+- Novel scores not commonly found elsewhere (e.g., [FIxed Risk Multicategorical (FIRM) score](https://doi.org/10.1002/qj.4266), [Flip-Flop Index](https://doi.org/10.1002/met.1732)).
+- Complex scores (e.g., threshold-weighted continuous ranked probability score (twCRPS))
+- Commonly-used scores are also included, so that `scores` can be used as a standalone package.
```
I modified the text in this dot point compared to the JOSS paper draft. In the JOSS paper draft it says "Commonly-used scores are also included, meeting user requests to use scores as a standalone package." If anyone would prefer the dot point on this page to be identical to the JOSS paper draft, let me know.

Steph-Chong commented on docs/key_features.md:
```diff
@@ -0,0 +1,26 @@
+## Usability
+
+- A companion Jupyter Notebook [tutorial](project:./tutorials/Tutorial_Gallery.md) for each metric and statistical test that demonstrates its use in practice.
+- Novel scores not commonly found elsewhere (e.g., FIxed Risk Multicategorical (FIRM) score ([Taggart et al., 2022](https://doi.org/10.1002/qj.4266)), Flip-Flop Index ([Griffiths et al., 2019](https://doi.org/10.1002/met.1732), [2021](https://doi.org/10.1071/ES21010))). 
```
Re. Flip-Flop Index link:
- Ideally I would link to the entry in the "included.md" table, but I don't know how to do that.
- Should it link to the paper reference (as is currently the case), or should it link to the API entry?

FIRM link:
- Ideally I would link to the entry in the "included.md" table, but I don't know how to do that.
- Should it link to the paper references (as is currently the case), or should it link to the API entry?

Steph-Chong replied:
@nicholasloveday can I confirm the correct term for FIRM is FIxed Risk Multicategorical (FIRM) Score?

I can't access the https://doi.org/10.1002/qj.4266 article, as it is paywalled.

But when I looked at the second version of the paper on arXiv (https://arxiv.org/pdf/2108.12814) - which I realise may not be identical to the final published version - there were two different descriptions of what FIRM stood for.

- On pages 3 and 7, it said "FIxed Risk Multicategory (FIRM) Forecast Framework"
- On page 15, it said "FIxed Risk Multicategorical (FIRM) framework"

So I just wanted to check that "FIxed Risk Multicategorical (FIRM) score" is correct (and that it should not be "FIxed Risk Multicategory (FIRM) score").

nicholasloveday replied:
@Steph-Chong I had to wait until today to chat to Rob. We can stick with "multicategorical".

Steph-Chong commented:
An alternative option re. the novel, complex +/- commonly-used scores dot points would be something along the lines of:

- [Over 50 metrics, statistical techniques and data processing tools](included.md), including:
  - Commonly-used metrics (e.g., mean absolute error (MAE), root mean square error (RMSE))
  - Novel scores not commonly found elsewhere (e.g., FIxed Risk Multicategorical (FIRM) score ([Taggart et al., 2022](https://doi.org/10.1002/qj.4266)), Flip-Flop Index ([Griffiths et al., 2019](https://doi.org/10.1002/met.1732), [2021](https://doi.org/10.1071/ES21010))).
  - Complex scores (e.g., threshold-weighted continuous ranked probability score (twCRPS))
  - Statistical tests (e.g., the Diebold-Mariano ([Diebold & Mariano, 1995](https://doi.org/10.3386/t0169)) test, with both
the Harvey et al. ([1997](https://doi.org/10.1016/S0169-2070(96)00719-4)) and Hering & Genton ([2011](https://doi.org/10.1198/tech.2011.10136)) modifications).

If the above option is preferred, (a) it could be left as is, (b) the dot points could be reordered, (c) some of the info could be removed so that it is shorter. But I wanted to put something along these lines forward for consideration.

tennlee commented:
> An alternative option re. the novel, commonly-used +/- complex scores dot points would be something along the lines of:
> 
> * [Over 50 metrics, statistical techniques and data processing tools](included.md), including:
>   
>   * Commonly-used metrics (e.g., mean absolute error (MAE), root mean squared error (RMSE))
>   * Novel scores not commonly found elsewhere (e.g., FIxed Risk Multicategorical (FIRM) score ([Taggart et al., 2022](https://doi.org/10.1002/qj.4266)), Flip-Flop Index ([Griffiths et al., 2019](https://doi.org/10.1002/met.1732), [2021](https://doi.org/10.1071/ES21010))).
>   * Complex scores (e.g., threshold-weighted continuous ranked probability score (twCRPS))
>   * Statistical tests (e.g., the Diebold-Mariano ([Diebold & Mariano, 1995](https://doi.org/10.3386/t0169)) test, with both the Harvey et al. ([1997](https://doi.org/10.1016/S0169-2070(96)00719-4)) and Hering & Genton ([2011](https://doi.org/10.1198/tech.2011.10136)) modifications).
> 
> If the above option is preferred, (a) it could be left as is, (b) the dot points could be reordered, (c) a dot point about data processing tools could be added, (d) some of the info could be removed so that it is shorter. But I wanted to put something along these lines forward for consideration.
> 
> Or perhaps, more along the lines of:
> 
> * [Over 50 metrics, statistical techniques and data processing tools](included.md), including:
>   
>   * Commonly-used scores (e.g., MAE, RMSE) and contigency table metrics (e.g., FAR, POD, POFD)
>   * Novel scores not commonly found elsewhere (e.g., [FIRM](https://doi.org/10.1002/qj.4266), [Flip-Flop Index](https://doi.org/10.1002/met.1732))
>   * Complex scores (e.g., twCRPS)
>   * Statistical tests (e.g., the [Diebold-Mariano](https://doi.org/10.3386/t0169) test, with both the Harvey et al. ([1997](https://doi.org/10.1016/S0169-2070(96)00719-4)) and Hering & Genton ([2011](https://doi.org/10.1198/tech.2011.10136)) modifications).
>   * Pre-processing tools for preparing data
> 
> Again, some of the dot points could be removed, or it can be modified. But I wanted to present some ideas.

For the docs (as opposed to the paper, which I think does not need this change), I think the sub-dot-point approach does make it easier to read. I personally prefer the version with more detail but could be easily swayed.

Steph-Chong commented:
> > An alternative option re. the novel, commonly-used +/- complex scores dot points would be something along the lines of:
> > 
> > * [Over 50 metrics, statistical techniques and data processing tools](included.md), including:
> >   
> >   * Commonly-used metrics (e.g., mean absolute error (MAE), root mean squared error (RMSE))
> >   * Novel scores not commonly found elsewhere (e.g., FIxed Risk Multicategorical (FIRM) score ([Taggart et al., 2022](https://doi.org/10.1002/qj.4266)), Flip-Flop Index ([Griffiths et al., 2019](https://doi.org/10.1002/met.1732), [2021](https://doi.org/10.1071/ES21010))).
> >   * Complex scores (e.g., threshold-weighted continuous ranked probability score (twCRPS))
> >   * Statistical tests (e.g., the Diebold-Mariano ([Diebold & Mariano, 1995](https://doi.org/10.3386/t0169)) test, with both the Harvey et al. ([1997](https://doi.org/10.1016/S0169-2070(96)00719-4)) and Hering & Genton ([2011](https://doi.org/10.1198/tech.2011.10136)) modifications).
> > 
> > If the above option is preferred, (a) it could be left as is, (b) the dot points could be reordered, (c) a dot point about data processing tools could be added, (d) some of the info could be removed so that it is shorter. But I wanted to put something along these lines forward for consideration.
> > Or perhaps, more along the lines of:
> > 
> > * [Over 50 metrics, statistical techniques and data processing tools](included.md), including:
> >   
> >   * Commonly-used scores (e.g., MAE, RMSE) and contigency table metrics (e.g., FAR, POD, POFD)
> >   * Novel scores not commonly found elsewhere (e.g., [FIRM](https://doi.org/10.1002/qj.4266), [Flip-Flop Index](https://doi.org/10.1002/met.1732))
> >   * Complex scores (e.g., twCRPS)
> >   * Statistical tests (e.g., the [Diebold-Mariano](https://doi.org/10.3386/t0169) test, with both the Harvey et al. ([1997](https://doi.org/10.1016/S0169-2070(96)00719-4)) and Hering & Genton ([2011](https://doi.org/10.1198/tech.2011.10136)) modifications).
> >   * Pre-processing tools for preparing data
> > 
> > Again, some of the dot points could be removed, or it can be modified. But I wanted to present some ideas.
> 
> For the docs (as opposed to the paper, which I think does not need this change), I think the sub-dot-point approach does make it easier to read. I personally prefer the version with more detail but could be easily swayed.

I have adjusted accordingly - the section now includes (a) the sub-dot-points and (b) the more detailed version. Feedback is welcome on the revised version.

(And yes, I agree the paper does not need to be changed, my suggestion was purely for the docs).

Steph-Chong commented on docs/key_features.md:
```diff
@@ -0,0 +1,28 @@
+  - commonly-used metrics (e.g., mean absolute error (MAE), root mean squared error (RMSE))
+  - novel scores not commonly found elsewhere (e.g., FIxed Risk Multicategorical (FIRM) score ([Taggart et al., 2022](https://doi.org/10.1002/qj.4266)), Flip-Flop Index ([Griffiths et al., 2019](https://doi.org/10.1002/met.1732), [2021](https://doi.org/10.1071/ES21010)))
+  - complex scores (e.g., threshold-weighted continuous ranked probability score (twCRPS))
+  - statistical tests (e.g., the Diebold-Mariano ([Diebold & Mariano, 1995](https://doi.org/10.3386/t0169)) test, with both the Harvey et al. ([1997](https://doi.org/10.1016/S0169-2070(96)00719-4)) and Hering & Genton ([2011](https://doi.org/10.1198/tech.2011.10136)) modifications).
```
(1) Should there also be a sub-dot-point about tools for pre-processing data?

(2), If yes, is there a preference for one of the following:
- Tools to pre-process data
- Tools for pre-processing data
- Tools for pre-processing data, including data matching, discretisation and cumulative density function manipulation
- Tools for pre-processing data in a variety of formats, including cumulative distribution functions (CDF).
- Something else?

Steph-Chong replied:
Note to self: if add tools to pre-process data sub-dot-point, make sure to remove full stop at end of proceeding dot point.

Steph-Chong replied:
If this sub-dot-point is included, I think I am leaning towards one of the following:
- tools for pre-processing data; including data matching, discretisation and cumulative density function manipulation.
- tools for pre-processing data - including data matching, discretisation and cumulative density function manipulation.

But I could be swayed in another direction, and am also open to alternative suggestions.

Steph-Chong replied:
Have raised this question in a new issue, so will resolve the conversation here.

Steph-Chong commented:
Is the fact `scores` contains isotonic regression a strong "selling point", such that it should be added as a dot point? I don't know the answer to this, so am happy to be guided by others.

Steph-Chong commented:
Are there any other major key features/"selling points" of `scores` that should be called out on this page?

Steph-Chong commented on docs/key_features.md:
```diff
@@ -0,0 +1,28 @@
+  - tabular, point, latitude/longitude or site-based data (e.g., forecasts for specific locations).
+- Handles missing data, masking of data and weighting of results.
+- Supports [xarray](https://xarray.dev/) datatypes, and works with [NetCDF4](https://doi.org/10.5065/D6H70CW6), [HDF5](https://github.com/HDFGroup/hdf5), [Zarr](https://zarr.dev) and [GRIB](https://codes.wmo.int/grib2) data sources among others. 
+
```
Should/could a dot point be added at the end of "Data Handling" section, along the lines of:
- [Some metrics](https://scores.readthedocs.io/en/228-documentation-testing-branch/included.html#pandas) work with [pandas](https://pandas.pydata.org/).

(I don't know if this actually falls under data handling or not).

If so, would remove the final dot point in the "Compatibility" section (which is about expanding support for pandas). If leaving the pandas dot point in the "compatibility" section, it might be worth changing the existing pandas dot point to the one suggested in this comment.

Steph-Chong replied:
Tennessee says leave this as is for now.

tennlee commented:
This is in great shape. I would like the remaining issues to be deal with as subsequent issues or PRs. I will merge this now since I think the current PR adds a lot of value to the documentation.
